src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs

```diff
@@ -245,9 +245,141 @@ void addArg(RefKind refKind, BoundExpression expression)
+                            var outputTemp = new BoundDagTemp(e.Syntax, e.SliceType, e);
+                            BoundExpression output = _tempAllocator.GetTemp(outputTemp);
+                            return _factory.AssignmentExpression(output, callExpr);
+                        }
+
                     default:
                         throw ExceptionUtilities.UnexpectedValue(evaluation);
                 }
+
+                BoundExpression makeImplicitIndexArgument(int index, BoundDagTemp lengthTemp)
+                {
+                    return index < 0
+                        ? _factory.IntSubtract(_tempAllocator.GetTemp(lengthTemp), _factory.Literal(-index))
+                        : _factory.Literal(index);
+                }
+
+                BoundExpression makeExplicitIndexArgument(int index, bool fromEnd)
+                {
+                    return _factory.New(
+                        WellKnownMember.System_Index__ctor,
+                        ImmutableArray.Create<BoundExpression>(
+                            _factory.Literal(Math.Abs(index)),
+                            _factory.Literal(fromEnd)));
+                }
+
+                BoundExpression makeExplicitRangeArgument(BoundDagSliceEvaluation e)
+                {
+                    return _factory.New(
+                        WellKnownMember.System_Range__ctor,
+                        ImmutableArray.Create(
+                            makeExplicitIndexArgument(e.StartIndex, fromEnd: false),
+                            makeExplicitIndexArgument(e.EndIndex, fromEnd: true)));
+                }
+
+                BoundExpression makeIndexerCall(SyntaxNode syntax, IndexerArgumentInfo info, PropertySymbol indexer, BoundExpression indexerArg)
+                {
+                    Debug.Assert(!info.ArgumentsOpt.IsDefaultOrEmpty);
+
+                    MethodSymbol getMethod = indexer.GetMethod;
+                    var firstArg = info.ArgumentsOpt[0];
+                    if (firstArg is BoundConversion conv)
+                        indexerArg = conv.UpdateOperand(indexerArg);
+
+                    ImmutableArray<RefKind> argumentRefKindsOpt = default;
+                    ImmutableArray<BoundExpression> args = _localRewriter.MakeArguments(
+                        syntax: syntax,
+                        rewrittenArguments: info.ArgumentsOpt.SetItem(0, indexerArg),
+                        methodOrIndexer: getMethod,
+                        expanded: info.Expanded,
+                        argsToParamsOpt: default,
+                        argumentRefKindsOpt: ref argumentRefKindsOpt,
+                        temps: out _,
+                        invokedAsExtensionMethod: false,
+                        enableCallerInfo: ThreeState.Unknown);
+                    return _factory.Call(input, getMethod, args);
+                }
+
+                bool isImplicit(PropertySymbol indexer)
```

Comment:
I'm not sure about this logic (seems odd to be checking for conversions here). It may be better to store the `IsImplicit` value into the bound node.